### DISCUSSION TITLE
MediatR.Contracts Question

### DISCUSSION THREAD

**Discussion by buvinghausen** [LuckyPennySoftware/MediatR#695] [Q&A]
Jimmy,

I like the direction of splitting the interfaces into the contracts package because that would let services depend on the contracts assembly and not the entire MediatR library which can be plugged in at the hosting level.

My question is why weren't all the interfaces moved over?  I use ISender in both my ApiControllerBase and AsyncValidatorBase and I use IRequestHandler<,> in my CommandHandlerBase class since it ultimately defers the commands to NServiceBus.

Buvy

**Comment by jbogard** [comment]
The other interfaces aren't nearly as important as the contract-associated ones, if I'm looking at what gets split apart in MediatR users. DTOs/contracts are often split out into separate packages/assemblies, but handlers? Not so much.

**Comment by ChristopherHaws** [comment]
I have a library at work that would benefit from adding the handler interfaces to the contracts package. We have a library that contains our DDD base types so that the projects in our department use the same base types. Right now the only dependency of the project is MediatR but we could make it rely on the contracts library if the handlers were in it. Here is an example of a base interface we have:

```c#
public interface IDomainEvent : INotification { }
public interface IDomainEventHandler<in TDomainEvent> : INotificationHandler<TDomainEvent> where TDomainEvent : IDomainEvent { }
```

**Comment by elementh** [comment]
Same as @ChristopherHaws for me, it's a must to have handler interfaces separeted in the contracts nuget.

Also, wouldn't it be nice to follow the not-so-offical-but-greatly-used convention and mark contract nugets as ".Abstractions"? It feels more at home with the rest of the dotnet world. It may be a silly request, but asking won't hurt!

**Comment by jbogard** [comment]
The point of the Contracts library is for cases where the request/response/notification DTOs are used *outside* of the process hosting the handlers. Blazor, Web APIs, etc. And in those cases the process *creating* the DTOs may also have its own reference to MediatR. And that version may not match. The Contracts library creates a common dependency only on what the message contracts need, and it's a dependency that will version much more slowly than MediatR itself.

As for the handler interfaces, those do actually change roughly at the same rate as MediatR. I've created and removed base classes, added parameters, changed parameters. If I broke out the interfaces into their own assembly, that assembly version would always be versioned and released when MediatR releases, so the net effect would be identical.

**Comment by buvinghausen** [comment]
@jbogard I fundamentally understand your point about versioning but the way I have architected my services each service sends all queries and commands via MediatR so they need access to the ISender interface (while I don't use IMediator or IPublisher I would imagine including them too).  I use default query handlers resolved via open generics using dry IOC so that particular group of handlers is bolted in at the hosting level but since command handlers can do some in process work before shipping the message to NServiceBus I need access to IRequestHandler<,>.  All the pipeline setup is done in the hosting application so no concerns there.

We deploy multiple services to a single host so the setup is like this

![Untitled Diagram drawio](https://user-images.githubusercontent.com/1130210/150206311-48df5ac5-db2b-481f-a8aa-2481f8603669.png)

**Reply by jbogard** [reply]
I don't understand what problem splitting the interfaces and base classes in a separate assembly would solve. There's almost nothing left in the MediatR assembly once you do that, just internal things, well-known behaviors, and the `Mediator` class itself.

**Comment by ChristopherHaws** [comment]
After thinking about this more, I think I agree with you @jbogard. Contracts was a good name for your package, it is a lot more accurate than calling it Abstractions. I think my library could depend on an abstractions package, but as you said they change at nearly the same rate as the rest of the code so there is no big benefit. The only benefit I can think of would be that dev's wouldn't be tempted to use a type they shouldn't be.

**Comment by fivage** [comment]
I have the issue that I can not call a IMediator.Send(...) from a net462 lib anymore, so that would also be a good reason to have this interface in the Contracts nuget.
